app/src/main/java/io/homeassistant/companion/android/nfc/TagReaderActivity.kt

```diff
@@ -0,0 +1,90 @@
+package io.homeassistant.companion.android.nfc
+
+import android.content.Intent
+import android.net.Uri
+import android.nfc.NdefMessage
+import android.nfc.NfcAdapter
+import android.os.Bundle
+import android.util.Log
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+import io.homeassistant.companion.android.R
+import io.homeassistant.companion.android.common.dagger.GraphComponentAccessor
+import io.homeassistant.companion.android.domain.integration.IntegrationUseCase
+import io.homeassistant.companion.android.util.UrlHandler
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+
+class TagReaderActivity : AppCompatActivity() {
+
+    val TAG = TagReaderActivity::class.simpleName
+
+    private val mainScope: CoroutineScope = CoroutineScope(Dispatchers.Main + Job())
+
+    @Inject
+    lateinit var integrationUseCase: IntegrationUseCase
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_tag_reader)
+
+        setSupportActionBar(findViewById(R.id.toolbar))
+
+        // Inject components
+        DaggerProviderComponent
+            .builder()
+            .appComponent((application as GraphComponentAccessor).appComponent)
+            .build()
+            .inject(this)
+
+        if (NfcAdapter.ACTION_NDEF_DISCOVERED == intent.action) {
+            val rawMessages = intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES)
+            val ndefMessage = rawMessages[0] as NdefMessage?
+            mainScope.launch {
+                try {
+                    val url = ndefMessage?.records?.get(0)?.toUri().toString()
+                    handleTag(url)
+                } catch (e: Exception) {
+                    val message = R.string.nfc_processing_tag_error
+                    Toast.makeText(this@TagReaderActivity, message, Toast.LENGTH_LONG).show()
+                    Log.e(TAG, e.message)
```

Comment:
We should add the whole exception and add a custom description.
`Log.e(TAG, "Unable to handle url: $url", e)`